velox/functions/prestosql/tests/CustomTypeCastTest.cpp

```diff
@@ -0,0 +1,215 @@
+            tryCast);
+
+    auto result = evaluate<SimpleVector<EvalType<TTo>>>(typedExpr, input);
+
+    assertEqualVectors(expected, result);
+  }
+
+  template <typename TFrom, typename TTo>
+  void testCustomCast(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      std::vector<std::optional<TFrom>> input,
+      std::vector<std::optional<TTo>> expected,
+      bool tryCast = false) {
+    evaluateCustomCast<TTo>(
+        fromType,
+        toType,
+        makeRowVector({makeNullableFlatVector<TFrom>(input)}),
+        makeNullableFlatVector<TTo>(expected),
+        tryCast);
+  }
+
+  template <typename TFrom, typename TTo>
+  void testCustomCastConstEncoding(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      std::vector<std::optional<TFrom>> input,
+      std::vector<std::optional<TTo>> expected,
+      bool tryCast = false) {
+    auto inputVector = makeNullableFlatVector<TFrom>(input);
+    auto constInput = BaseVector::wrapInConstant(5, 0, inputVector);
+
+    auto expectedVector = makeNullableFlatVector<TTo>(expected);
+    auto constExpected = BaseVector::wrapInConstant(5, 0, expectedVector);
+
+    evaluateCustomCast<TTo>(
+        fromType, toType, makeRowVector({constInput}), constExpected, tryCast);
+  }
+
+  template <typename TFrom, typename TTo>
+  void testCustomCastDictEncoding(
+      const TypePtr& fromType,
+      const TypePtr& toType,
+      std::vector<std::optional<TFrom>> input,
+      std::vector<std::optional<TTo>> expected,
+      bool tryCast = false) {
+    std::shared_ptr<const core::ITypedExpr> inputField =
+        std::make_shared<const core::FieldAccessTypedExpr>(fromType, "c0");
+    std::shared_ptr<const core::ITypedExpr> callExpr =
+        std::make_shared<const core::CallTypedExpr>(
+            fromType,
+            std::vector<std::shared_ptr<const core::ITypedExpr>>{inputField},
+            "testing_dictionary");
+    std::shared_ptr<const core::ITypedExpr> typedExpr =
+        std::make_shared<const core::CastTypedExpr>(
+            toType,
+            std::vector<std::shared_ptr<const core::ITypedExpr>>{callExpr},
+            tryCast);
+
+    auto inputVector = makeRowVector({makeNullableFlatVector<TFrom>(input)});
+
+    auto expectedVector = makeNullableFlatVector<TTo>(expected);
+    auto indices = ::makeIndicesInReverse(inputVector->size(), pool());
+
+    auto result = evaluate<SimpleVector<EvalType<TTo>>>(typedExpr, inputVector);
+    assertEqualVectors(
+        wrapInDictionary(indices, inputVector->size(), expectedVector), result);
+  }
+};
+
+TEST_F(CastExprTest, testJsonCustomCastInteger) {
+  testCustomCast<int64_t, Json>(
+      BIGINT(),
+      JSON(),
+      {1, -3, INT64_MAX, INT64_MIN, std::nullopt},
+      {"1"_sv,
+       "-3"_sv,
+       "9223372036854775807"_sv,
+       "-9223372036854775808"_sv,
+       std::nullopt},
+      false);
+
+  testCustomCastConstEncoding<int64_t, Json>(
```

Comment:
It might be cleaner to modify testCast to test flat, constant and dictionary encodings and not require 3 separate calls here.

Reply:
Thank you for this suggestion! I modified `testCast()` to test with flat encoding, constant encoding, and dictionary encoding with the same inputs.